Add Cell Dissemination via Partial Message Specification (#4558)
ethereum/consensus-specs: specs/fulu/p2p-interface.md

    kzg_proofs: List[KZGProof, MAX_BLOB_COMMITMENTS_PER_BLOCK]
    # Optional header, only sent on eager pushes
    header: List[PartialDataColumnHeader, 1]
```

#### `PartialDataColumnPartsMetadata`

Peers communicate the cells available with a bitmap. A set bit (`1`) at index
`i` means that the peer has the cell at index `i`. Peers explicitly request
cells with a second request bitmap of the same length that is set to `1` if the
peer would like to receive or provide this cell.

This is encoded as the following SSZ container:

```python
class PartialDataColumnPartsMetadata(Container):
    available: Bitlist[MAX_BLOB_COMMITMENTS_PER_BLOCK]
    requests: Bitlist[MAX_BLOB_COMMITMENTS_PER_BLOCK]
```

This means that for each cell there are two bits of state. Where the first bit
represents the bit from the available bitlist, and the second bit represents the
bit from the requests bit.

| Bits | Description                                          |
| :--: | ---------------------------------------------------- |
|  00  | The peer does not have the cell and does not want it |
|  01  | The peer does not have the cell and does want it     |

> |  1X  | The peer has the cell and is willing to provide it   |

Author: Currently it's common for clients to just set the request bitmask to all 1's. If they don't change that when they receive a cell it'll be common to see `11`.

The argument for this change was that semantically a request bit doesn't make sense if the cell is available, so we should ignore the request bit if the available bit is set. [Previously](https://github.com/ethereum/consensus-specs/pull/4558/changes/876d1c3a9d3f47f557789a6bc6e57e932e7da633) we treated `10` as unused and ignored, but after discussion with @sukunrt and @raulk, we opted for ignoring the request bit and treating `11` and `10` the same.

> |  1X  | The peer has the cell and is willing to provide it   |

Reviewer: Got it. Okay this makes sense. Thank you for explaining 🙂

##### `PartialDataColumnSidecar`

The `PartialDataColumnSidecar `is identical to the `DataColumnSidecar `container
with the exception that fields and cells may be None.

Some fields are only set when a peer is eagerly pushing this container.
Otherwise, if a peer is sending the container as a response, there is no need to
include redundant fields.

```python
class PartialDataColumnSidecar(Container):
    signed_block_header: SignedBeaconBlockHeader
    cells_present_bitmap: Bitlist[MAX_BLOB_COMMITMENTS_PER_BLOCK]
    column: List[Cell, MAX_BLOB_COMMITMENTS_PER_BLOCK]
    kzg_proofs: List[KZGProof, MAX_BLOB_COMMITMENTS_PER_BLOCK]
```

##### Partial IHAVE/IWANT

Peers communicate IHAVE and IWANTS succintly with a bitmap. For IHAVE, a set bit (`1`)
means that the peer has the corresponding cell. For IWANT, a set bit means the
peer wants the corresponding cell. The bitmap is encoded by the following function:

```python
def create_bitmap(set_indices, n):
    bitmap = [0] * ((n + 7) // 8)
    for i in set_indices:
        bitmap[i // 8] |= 1 << (i % 8)
    return bitmap
```

For example, if the peer wanted cells at positions 0,3,9, the corresponding

> IWANT bitmap would be: `0000 1001 000 0010`.

Reviewer 2: What's `n` in the example?
000 -> 0000 , and also the byte order looks wrong, shouldn't we have MSB-first?

Author: thanks for the 000 -> 0000 correction.

n can be anything from 9 to 16 in this case, but I should just say 12 to be clear in the example.

> The byte order looks wrong, shouldn't we have MSB-first?

https://www.online-python.com/ietUIq50Ol

    """
    Verify if the KZG proofs are correct.
    """
    # The column index also represents the cell index
    cell_indices = [i for i, b in enumerate(bin(bitmap)[:1:-1]) if b == '1']

    # Batch verify that the cells match the corresponding commitments and proofs
    return verify_cell_kzg_proof_batch(
        commitments_bytes=[all_commitments[i] for i in cell_indices],
        cell_indices=cell_indices,
        cells=sidecar.column,
        proofs_bytes=sidecar.kzg_proofs,
    )
```

##### Eager Pushing

In contrast to standard Gossipsub, A Client usually requests missing parts from
a peer. A client can send its partial IWANT before receiving a peer's partial
IHAVE to register interest in certain parts. This can introduce extra latency
compared to a Client unconditionally pushing messages to a peer.

To address this tradeoff a Client may choose to eagerly push some (or all) of
the cells it has. Clients SHOULD only do this when they are reasonably confident
that a peer does not have the provided cells. For example, a proposer including

> private blobs SHOULD eagerly push the cells corresponding to the private blobs.

Author: I don't see how this is worse than the current status quo. Currently we always eagerly push a whole column _unless_ we receive an IDONTWANT.

This is the same except an IDONTWANT is replaced by an updated partial IHAVE.

> private blobs SHOULD eagerly push the cells corresponding to the private blobs.

Reviewer 3: There's a subtle difference though. Before, your eager push was guaranteed to be useful if the peer was lacking _any_ blob. Now, you can push data that's knowingly useless (e.g. an attacker always pushes a partial cells they know their peer has based on EL tx announcements from the same IP). At 72 blobs, for a supernode that's only lacking 1 blob, this path exposes it to receiving 18MiB x D of knowingly redundant data, when all it needed was 256KiB (1 row x 128 columns x 2KiB cell size).

- Update gossip rule related data structures (i.e. update the anti-equivocation
  cache).

#### Partial Columns

Gossipsub's [Partial Message
Extension](https://github.com/libp2p/specs/pull/685) enables exchanging
selective parts of a message rather than the whole. The specification here
describes how Consensus Clients use Partial Messages to disseminate cells.

*Editor's Note*: This change MUST NOT be merged before the linked libp2p/specs.

##### `PartialDataColumnSidecar`

The `PartialDataColumnSidecar` is similar to the `DataColumnSidecar `container,
except that only the cells and proofs identified by the bitmap are present.

```python
class PartialDataColumnSidecar(Container):
    cells_present_bitmap: Bitlist[MAX_BLOB_COMMITMENTS_PER_BLOCK]
    partial_column: List[Cell, MAX_BLOB_COMMITMENTS_PER_BLOCK]
    kzg_proofs: List[KZGProof, MAX_BLOB_COMMITMENTS_PER_BLOCK]
```

> 

Reviewer 4: Seems necessary for verification

```suggestion
```python
class PartialDataColumnSidecar(Container):
    cells_present_bitmap: Bitlist[MAX_BLOB_COMMITMENTS_PER_BLOCK]
    partial_column: List[Cell, MAX_BLOB_COMMITMENTS_PER_BLOCK]
    kzg_proofs: List[KZGProof, MAX_BLOB_COMMITMENTS_PER_BLOCK]
    beacon_block_root: Root
```
```

*Note*: The `PartialDataColumnSidecar` can not be validated until the
corresponding block with the commitments is available. Implementations SHOULD
maintain a bounded number (0 is a valid number) of messages until the block is
available and the messages can be validated.

##### Parts Metadata

Peers communicate the cells available with a bitmap. A set bit (`1`) at index `i`
means that the peer has cell the cell at index `i`. The bitmap is encoded as a
Bitlist.

##### Encoding and Decoding Responses

All responses should be encoded and decoded with the PartialDataColumnSidecar
container.

##### Validation

TODO add full validation rules

######  `verify_partial_data_column_sidecar_kzg_proofs`

```python
def verify_partial_data_column_sidecar_kzg_proofs(sidecar: PartialDataColumnSidecar, all_commitments: List[KZGCommitment, MAX_BLOB_COMMITMENTS_PER_BLOCK]) -> bool:
    """
    Verify if the KZG proofs are correct.
    """
    # The column index also represents the cell index

>     cell_indices = [i for i, b in enumerate(bin(bitmap)[:1:-1]) if b == '1']

Reviewer 4: ```suggestion
    cell_indices = [i for i, b in enumerate(sidecar.cells_present_bitmap) if b]
```

*Editor's Note*: This change MUST NOT be merged before the linked libp2p/specs.

##### Partial Message Group ID

When sending a partial message, the Gossipsub Group ID MUST be the block root.

##### `PartialDataColumnSidecar`

The `PartialDataColumnSidecar` is similar to the `DataColumnSidecar` container,
except that only the cells and proofs identified by the bitmap are present.

The column index is inferred from the subnet (Gossipsub topic).

```python
class PartialDataColumnSidecar(Container):
    cells_present_bitmap: Bitlist[MAX_BLOB_COMMITMENTS_PER_BLOCK]
    partial_column: List[Cell, MAX_BLOB_COMMITMENTS_PER_BLOCK]
    kzg_proofs: List[KZGProof, MAX_BLOB_COMMITMENTS_PER_BLOCK]
    # Optional header, only sent on eager pushes

>     header: List[PartialDataColumnHeader, 1]

Author: > I assume the primary motivation of not using `Union[None, PartialDataColumnHeader]` is missing support in some SSZ implementations, and the fact that it is not currently used elsewhere in Ethereum?

Correct. See also this PR: https://github.com/ethereum/consensus-specs/pull/3906

##### Validation

Validating partial messages happens in two parts. First the
`PartialDataColumnHeader` needs to be validated, then the cell and proof data.

Once a `PartialDataColumnHeader` is validated for a corresponding block
on any subnet (gossipsub topic), it can be used for all subnets.

Due to the nature of partial messages, it's possible to get the
`PartialDataColumnHeader` with no cells, and get cells in a future response.

For all partial messages:
- _[IGNORE]_ If the received partial message contains only cell data, the

>   node has not yet seen the corresponding `PartialDataColumnHeader`.

Reviewer 5: Do you mean

```
- _[IGNORE]_ If the received partial message contains only cell data and the
  node has not yet seen the corresponding `PartialDataColumnHeader`.
```

Author: Sorry, I always get tripped up with the phrasing of these rules.

I mean to say If the message only contains cell data and we haven't seen the header yet we should ignore it.

Which, with the way the rules are formatted (`<action> <expectation>`, and only do `action` if expectation is not met), I believe it should be:

```
If the received partial message contains only cell data, the node has seen the corresponding `PartialDataColumnHeader`.
```

So the quoted version is incorrect (It has an extra "not")

  block -- i.e.
  `get_checkpoint_block(store, block_header.parent_root, store.finalized_checkpoint.epoch) == store.finalized_checkpoint.root`.
- _[REJECT]_ The header's `kzg_commitments` field inclusion proof is valid as
  verified by `verify_partial_data_column_header_inclusion_proof`.
- _[REJECT]_ The header is proposed by the expected `proposer_index` for the
  block's slot in the context of the current shuffling (defined by
  `block_header.parent_root`/`block_header.slot`). If the `proposer_index`
  cannot immediately be verified against the expected shuffling, the header MAY
  be queued for later processing while proposers for the block's branch are
  calculated -- in such a case _do not_ `REJECT`, instead `IGNORE` this message.

For verifying the cells in a partial message:

- _[IGNORE]_ If the received partial message contains only cell data, the node
  has seen a valid corresponding `PartialDataColumnHeader`.
- _[IGNORE]_ The corresponding header is not from a future slot. See related
  header check above for more details
- _[IGNORE]_ The corresponding header is from a slot greater than the latest
  finalized slot. See related header check above for more details

> - _[REJECT]_ The cells present bitmap length is equal to the number of KZG

Reviewer: ```suggestion
- _[IGNORE]_ The corresponding header is not from a future slot. See related
  header check above for more details.
- _[IGNORE]_ The corresponding header is from a slot greater than the latest
  finalized slot. See related header check above for more details.
- _[REJECT]_ The cells present bitmap length is equal to the number of KZG
```

metadata. This registers interest in certain parts, even if the peer does not
have these parts yet.

This request can introduce extra latency compared to a peer unconditionally
pushing messages, especially in the first hop of dissemination.

To address this tradeoff, a client MAY choose to eagerly push some (or all) of
the cells it has. Clients SHOULD only do this when they are reasonably confident
that a peer does not have the provided cells. For example, a proposer including
private blobs SHOULD eagerly push the cells corresponding to the private blobs.

Clients, by default, SHOULD NOT eagerly push cells when proposing a block.
Clients SHOULD expose a flag to opt-in to eagerly pushing all cells when
proposing a block.

Clients SHOULD eagerly push the `PartialDataColumnHeader` to inform peers as to
which blobs are included in this block, and therefore which cells they are
missing. Clients SHOULD NOT send a `PartialDataColumnHeader` non-eagerly, as
this is wasted bandwidth.

Clients MAY choose to not eagerly push the `PartialDataColumnHeader` if it has
previously sent the header to the peer on another topic.

> Clients SHOULD request cell data from peers after validating a

Reviewer: Just want to mention some inconsistent language for:

* cells
* cell/proof data
* cell data

I'm pretty sure you always mean "cells and proofs" in these. I would suggest improving this.

- Update gossip rule related data structures (i.e. update the anti-equivocation
  cache).

###### Partial Messages on `data_column_sidecar_{subnet_id}`

Validating partial messages happens in two parts. First, the
`PartialDataColumnHeader` needs to be validated, then the cell and proof data.

Once a `PartialDataColumnHeader` is validated for a corresponding block on any
subnet (gossipsub topic), it can be used for all subnets.

Due to the nature of partial messages, it is possible to get the
`PartialDataColumnHeader` with no cells, and get cells in a future response.

For all partial messages:

- _[IGNORE]_ If the received partial message contains only cell data, the node
  has seen the corresponding `PartialDataColumnHeader`.

For verifying the `PartialDataColumnHeader` in a partial message:

> - _[IGNORE]_ The header is the first valid header for the given block root.

Author: should be covered by 48cd1f9f

container.

##### Encoding and decoding responses

All responses MUST be encoded and decoded with the `PartialDataColumnSidecar`
container.

##### Eager pushing

In contrast to standard gossipsub, a client explicitly requests missing parts
from a peer. A client can send its request before receiving a peer's parts
metadata. This registers interest in certain parts, even if the peer does not
have these parts yet.

This request can introduce extra latency compared to a peer unconditionally
pushing messages, especially in the first hop of dissemination.

To address this tradeoff, a client MAY choose to eagerly push some (or all) of
the cells it has. Clients SHOULD only do this when they are reasonably confident
that a peer does not have the provided cells. For example, a proposer including
private blobs SHOULD eagerly push the cells corresponding to the private blobs.

Clients SHOULD eagerly push the `PartialDataColumnHeader` to inform peers as to
which blobs are included in this block, and therefore which cells they are
missing. Clients SHOULD NOT send a `PartialDataColumnHeader` non-eagerly, as
this is wasted bandwidth.

> Clients MAY choose to not eagerly push the `PartialDataColumnHeader` if it has

Reviewer 6: Okay so we're not enforcing that the header should be sent ONLY once per block ? What do you think is a valid use case for allowing this ?

Author: this is a `MAY`. If you have the same peer for multiple topics, you don't need to send them duplicate headers on all topics.

pushing messages, especially in the first hop of dissemination.

To address this tradeoff a client MAY choose to eagerly push some (or all) of
the cells it has. Clients SHOULD only do this when they are reasonably confident
that a peer does not have the provided cells. For example, a proposer including
private blobs SHOULD eagerly push the cells corresponding to the private blobs.

Clients SHOULD eagerly push the `PartialDataColumnHeader` to inform peers as to
which blobs are included in this block, and therefore which cells they are

> missing. Clients SHOULD NOT send a `PartialDataColumnHeader` non-eagerly, as

Reviewer 5: Are we including some sentence to indicate that nodes can choose to not push the header eagerly to a node if it already has done so in another message?
